File: sdk/typescript/src/services/agent.ts
/**
 * ACGS-2 Agent Service
 * Constitutional Hash: cdd01ef066bc6cf2
 */

import EventEmitter from 'eventemitter3';
import { ACGS2Client } from '../client';
import {
  CONSTITUTIONAL_HASH,
  AgentMessage,
  AgentMessageSchema,
  MessageType,
  Priority,
  SendMessageRequest,
  PaginationParams,
  PaginatedResponse,
  ACGS2Event,
  EventHandler,
} from '../types';
import { generateUUID, nowISO, ValidationError, Logger, createLogger, silentLogger } from '../utils';

// =============================================================================
// Agent Types
// =============================================================================

export interface AgentInfo {
  id: string;
  name: string;
  type: string;
  status: 'active' | 'inactive' | 'suspended';
  capabilities: string[];
  metadata: Record<string, string>;
  lastSeen: string;
  constitutionalHash: string;
}

export interface AgentRegistration {
  name: string;
  type: string;
  capabilities: string[];
  metadata?: Record<string, string>;
}

// =============================================================================
// Agent Service
// =============================================================================

export class AgentService {
  private readonly basePath = '/api/v1/agents';
  private readonly eventEmitter: EventEmitter;
  private readonly logger: Logger;
  private agentId?: string;
  private websocket?: WebSocket;
  private reconnectAttempts = 0;
  private readonly maxReconnectAttempts = 5;

  constructor(
    private readonly client: ACGS2Client,
    options?: { logger?: Logger }
  ) {
    this.eventEmitter = new EventEmitter();
    this.logger = options?.logger ?? silentLogger;
  }

  // ===========================================================================
  // Agent Registration
  // ===========================================================================

  /**
   * Registers an agent with the ACGS-2 system
   */
  async register(registration: AgentRegistration): Promise<AgentInfo> {
    const response = await this.client.post<AgentInfo>(`${this.basePath}/register`, {
      ...registration,
      constitutionalHash: CONSTITUTIONAL_HASH,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to register agent');
    }

    this.agentId = response.data.id;
    this.logger.info(`Agent registered: ${this.agentId}`);
    return response.data;
  }

  /**
   * Unregisters the current agent
   */
  async unregister(): Promise<void> {
    if (!this.agentId) {
      throw new ValidationError('Agent not registered');
    }

    await this.client.delete(`${this.basePath}/${this.agentId}`);
    this.logger.info(`Agent unregistered: ${this.agentId}`);
    this.agentId = undefined;
  }

  /**
   * Gets the current agent's ID
   */
  getAgentId(): string | undefined {
    return this.agentId;
  }

  /**
   * Sets the agent ID (for existing agents)
   */
  setAgentId(agentId: string): void {
    this.agentId = agentId;
  }

  // ===========================================================================
  // Agent Information
  // ===========================================================================

  /**
   * Gets agent information by ID
   */
  async get(agentId: string): Promise<AgentInfo> {
    const response = await this.client.get<AgentInfo>(`${this.basePath}/${agentId}`);

    if (!response.success || !response.data) {
      throw new ValidationError(`Agent not found: ${agentId}`);
    }

    return response.data;
  }

  /**
   * Lists all registered agents
   */
  async list(params?: PaginationParams & {
    type?: string;
    status?: 'active' | 'inactive' | 'suspended';
  }): Promise<PaginatedResponse<AgentInfo>> {
    const response = await this.client.get<PaginatedResponse<AgentInfo>>(this.basePath, params);

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to list agents');
    }

    return response.data;
  }

  /**
   * Updates agent metadata
   */
  async updateMetadata(metadata: Record<string, string>): Promise<AgentInfo> {
    if (!this.agentId) {
      throw new ValidationError('Agent not registered');
    }

    const response = await this.client.patch<AgentInfo>(`${this.basePath}/${this.agentId}`, {
      metadata,
    });

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to update agent metadata');
    }

    return response.data;
  }

  /**
   * Sends a heartbeat to keep agent active
   */
  async heartbeat(): Promise<void> {
    if (!this.agentId) {
      throw new ValidationError('Agent not registered');
    }

    await this.client.post(`${this.basePath}/${this.agentId}/heartbeat`, {
      timestamp: nowISO(),
      constitutionalHash: CONSTITUTIONAL_HASH,
    });
  }

  // ===========================================================================
  // Messaging
  // ===========================================================================

  /**
   * Sends a message to another agent or broadcast
   */
  async sendMessage(request: SendMessageRequest): Promise<AgentMessage> {
    if (!this.agentId) {
      throw new ValidationError('Agent not registered');
    }

    const message: Omit<AgentMessage, 'id'> = {
      type: request.type,
      priority: request.priority ?? Priority.NORMAL,
      sourceAgentId: this.agentId,
      targetAgentId: request.targetAgentId,
      payload: request.payload,
      timestamp: nowISO(),
      correlationId: request.correlationId,
      constitutionalHash: CONSTITUTIONAL_HASH,
      metadata: request.metadata,
    };

    const response = await this.client.post<AgentMessage>(
      `${this.basePath}/${this.agentId}/messages`,
      message
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to send message');
    }

    return this.validateMessage(response.data);
  }

  /**
   * Sends a command to an agent
   */
  async sendCommand(
    targetAgentId: string,
    command: string,
    params?: Record<string, unknown>,
    options?: { priority?: Priority; correlationId?: string }
  ): Promise<AgentMessage> {
    return this.sendMessage({
      type: MessageType.COMMAND,
      priority: options?.priority ?? Priority.NORMAL,
      targetAgentId,
      payload: { command, params },
      correlationId: options?.correlationId,
    });
  }

  /**
   * Sends a query to an agent
   */
  async sendQuery(
    targetAgentId: string,
    query: string,
    params?: Record<string, unknown>,
    options?: { priority?: Priority; correlationId?: string }
  ): Promise<AgentMessage> {
    return this.sendMessage({
      type: MessageType.QUERY,
      priority: options?.priority ?? Priority.NORMAL,
      targetAgentId,
      payload: { query, params },
      correlationId: options?.correlationId,
    });
  }

  /**
   * Broadcasts an event to all agents
   */
  async broadcastEvent(
    eventType: string,
    data: Record<string, unknown>,
    options?: { priority?: Priority }
  ): Promise<AgentMessage> {
    return this.sendMessage({
      type: MessageType.EVENT,
      priority: options?.priority ?? Priority.NORMAL,
      payload: { eventType, data },
    });
  }

  /**
   * Gets messages for the current agent
   */
  async getMessages(params?: PaginationParams & {
    type?: MessageType;
    since?: string;
    unreadOnly?: boolean;
  }): Promise<PaginatedResponse<AgentMessage>> {
    if (!this.agentId) {
      throw new ValidationError('Agent not registered');
    }

    const response = await this.client.get<PaginatedResponse<AgentMessage>>(
      `${this.basePath}/${this.agentId}/messages`,
      params
    );

    if (!response.success || !response.data) {
      throw new ValidationError('Failed to get messages');
    }

    // Validate each message
    response.data.data = response.data.data.map((msg) => this.validateMessage(msg));
    return response.data;
  }

  /**
   * Acknowledges receipt of a message
   */
  async acknowledgeMessage(messageId: string): Promise<void> {
    if (!this.agentId) {
      throw new ValidationError('Agent not registered');
    }

    await this.client.post(`${this.basePath}/${this.agentId}/messages/${messageId}/ack`);
  }

  // ===========================================================================
  // Real-time Subscriptions
  // ===========================================================================

  /**
   * Subscribes to real-time messages via WebSocket
   */
  async subscribe(wsUrl: string): Promise<void> {
    if (this.websocket) {
      this.websocket.close();
    }

    return new Promise((resolve, reject) => {
      const url = `${wsUrl}?agentId=${this.agentId}&hash=${CONSTITUTIONAL_HASH}`;
      this.websocket = new WebSocket(url);

      this.websocket.onopen = () => {
        this.logger.info('WebSocket connected');
        this.reconnectAttempts = 0;
        resolve();
      };

      this.websocket.onmessage = (event) => {
        try {
          const data = JSON.parse(event.data);
          const message = this.validateMessage(data);
          this.eventEmitter.emit('message', message);
          this.eventEmitter.emit(`message:${message.type}`, message);
        } catch (error) {
          this.logger.error('Failed to parse WebSocket message:', error);
        }
      };

      this.websocket.onerror = (error) => {
        this.logger.error('WebSocket error:', error);
        this.eventEmitter.emit('error', error);
      };

      this.websocket.onclose = () => {
        this.logger.info('WebSocket disconnected');
        this.eventEmitter.emit('disconnected');
        this.attemptReconnect(wsUrl);
      };
    });
  }

  /**
   * Unsubscribes from real-time messages
   */
  unsubscribe(): void {
    if (this.websocket) {
      this.websocket.close();
      this.websocket = undefined;
    }
    this.reconnectAttempts = this.maxReconnectAttempts; // Prevent reconnection
  }

  /**
   * Registers a message handler
   */
  onMessage(handler: EventHandler<AgentMessage>): () => void {
    this.eventEmitter.on('message', handler);
    return () => this.eventEmitter.off('message', handler);
  }

  /**
   * Registers a handler for specific message types
   */
  onMessageType(type: MessageType, handler: EventHandler<AgentMessage>): () => void {
    this.eventEmitter.on(`message:${type}`, handler);
    return () => this.eventEmitter.off(`message:${type}`, handler);
  }

  /**
   * Registers a handler for connection errors
   */
  onError(handler: (error: Error) => void): () => void {
    this.eventEmitter.on('error', handler);
    return () => this.eventEmitter.off('error', handler);
  }

  /**
   * Registers a handler for disconnection
   */
  onDisconnected(handler: () => void): () => void {
    this.eventEmitter.on('disconnected', handler);
    return () => this.eventEmitter.off('disconnected', handler);
  }

  private async attemptReconnect(wsUrl: string): Promise<void> {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      this.logger.error('Max reconnection attempts reached');
      return;
    }

    this.reconnectAttempts++;
    const delay = Math.min(1000 * Math.pow(2, this.reconnectAttempts), 30000);
    this.logger.info(`Attempting reconnection in ${delay}ms (attempt ${this.reconnectAttempts})`);

    await new Promise((resolve) => setTimeout(resolve, delay));

    try {
      await this.subscribe(wsUrl);
    } catch (error) {
      this.logger.error('Reconnection failed:', error);
    }
  }

  // ===========================================================================
  // Helpers
  // ===========================================================================

  private validateMessage(data: unknown): AgentMessage {
    const result = AgentMessageSchema.safeParse(data);
    if (!result.success) {
      throw new ValidationError('Invalid message data', {
        validation: result.error.errors.map((e) => e.message),
      });
    }
    return result.data;
  }
}

File: src/core/sdk/python/acgs2_sdk/services/agent.py
"""
ACGS-2 Agent Service
Constitutional Hash: cdd01ef066bc6cf2
"""

from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any

try:
    from src.core.shared.types import JSONDict, JSONValue
except ImportError:
    JSONValue = str | int | float | bool | None | dict[str, Any] | list[Any]  # type: ignore[misc]
    JSONDict = dict[str, JSONValue]  # type: ignore[misc]

from acgs2_sdk.constants import AGENTS_ENDPOINT, CONSTITUTIONAL_HASH
from acgs2_sdk.models import (
    AgentInfo,
    AgentMessage,
    MessageType,
    PaginatedResponse,
    Priority,
    SendMessageRequest,
)

if TYPE_CHECKING:
    from acgs2_sdk.client import ACGS2Client


class AgentService:
    """Service for managing agents and messaging."""

    def __init__(self, client: "ACGS2Client") -> None:
        """Initialize the agent service.

        Args:
            client: ACGS2 client instance
        """
        self._client = client
        self._base_path = AGENTS_ENDPOINT
        self._agent_id: str | None = None

    @property
    def agent_id(self) -> str | None:
        """Get the current agent ID."""
        return self._agent_id

    @agent_id.setter
    def agent_id(self, value: str) -> None:
        """Set the agent ID."""
        self._agent_id = value

    async def register(
        self,
        name: str,
        agent_type: str,
        capabilities: list[str],
        metadata: dict[str, str] | None = None,
    ) -> AgentInfo:
        """Register a new agent.

        Args:
            name: Agent name
            agent_type: Agent type
            capabilities: Agent capabilities
            metadata: Optional metadata

        Returns:
            Registered agent info
        """
        data = await self._client.post(
            f"{self._base_path}/register",
            json={
                "name": name,
                "type": agent_type,
                "capabilities": capabilities,
                "metadata": metadata or {},
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        agent = AgentInfo.model_validate(data.get("data", data))
        self._agent_id = agent.id
        return agent

    async def unregister(self) -> None:
        """Unregister the current agent."""
        if not self._agent_id:
            raise ValueError("Agent not registered")
        await self._client.delete(f"{self._base_path}/{self._agent_id}")
        self._agent_id = None

    async def get(self, agent_id: str) -> AgentInfo:
        """Get agent info by ID.

        Args:
            agent_id: Agent ID

        Returns:
            Agent info
        """
        data = await self._client.get(f"{self._base_path}/{agent_id}")
        return AgentInfo.model_validate(data.get("data", data))

    async def list(
        self,
        page: int = 1,
        page_size: int = 50,
        agent_type: str | None = None,
        status: str | None = None,
    ) -> PaginatedResponse[AgentInfo]:
        """List registered agents.

        Args:
            page: Page number
            page_size: Items per page
            agent_type: Filter by type
            status: Filter by status

        Returns:
            Paginated list of agents
        """
        params: dict[str, Any] = {"page": page, "pageSize": page_size}
        if agent_type:
            params["type"] = agent_type
        if status:
            params["status"] = status

        data = await self._client.get(self._base_path, params=params)
        response_data = data.get("data", data)
        return PaginatedResponse[AgentInfo](
            data=[AgentInfo.model_validate(a) for a in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def heartbeat(self) -> None:
        """Send a heartbeat to keep agent active."""
        if not self._agent_id:
            raise ValueError("Agent not registered")
        await self._client.post(
            f"{self._base_path}/{self._agent_id}/heartbeat",
            json={
                "timestamp": datetime.now(UTC).isoformat(),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )

    async def send_message(self, request: SendMessageRequest) -> AgentMessage:
        """Send a message to another agent.

        Args:
            request: Message request

        Returns:
            Sent message
        """
        if not self._agent_id:
            raise ValueError("Agent not registered")

        data = await self._client.post(
            f"{self._base_path}/{self._agent_id}/messages",
            json={
                **request.model_dump(by_alias=True, exclude_none=True),
                "sourceAgentId": self._agent_id,
                "timestamp": datetime.now(UTC).isoformat(),
                "constitutionalHash": CONSTITUTIONAL_HASH,
            },
        )
        return AgentMessage.model_validate(data.get("data", data))

    async def send_command(
        self,
        target_agent_id: str,
        command: str,
        params: dict[str, Any] | None = None,
        priority: Priority = Priority.NORMAL,
    ) -> AgentMessage:
        """Send a command to an agent.

        Args:
            target_agent_id: Target agent ID
            command: Command name
            params: Command parameters
            priority: Message priority

        Returns:
            Sent message
        """
        return await self.send_message(
            SendMessageRequest(
                type=MessageType.COMMAND,
                priority=priority,
                target_agent_id=target_agent_id,
                payload={"command": command, "params": params or {}},
            )
        )

    async def send_query(
        self,
        target_agent_id: str,
        query: str,
        params: dict[str, Any] | None = None,
        priority: Priority = Priority.NORMAL,
    ) -> AgentMessage:
        """Send a query to an agent.

        Args:
            target_agent_id: Target agent ID
            query: Query string
            params: Query parameters
            priority: Message priority

        Returns:
            Sent message
        """
        return await self.send_message(
            SendMessageRequest(
                type=MessageType.QUERY,
                priority=priority,
                target_agent_id=target_agent_id,
                payload={"query": query, "params": params or {}},
            )
        )

    async def broadcast_event(
        self,
        event_type: str,
        data: JSONDict,
        priority: Priority = Priority.NORMAL,
    ) -> AgentMessage:
        """Broadcast an event to all agents.

        Args:
            event_type: Event type
            data: Event data
            priority: Message priority

        Returns:
            Sent message
        """
        return await self.send_message(
            SendMessageRequest(
                type=MessageType.EVENT,
                priority=priority,
                payload={"eventType": event_type, "data": data},
            )
        )

    async def get_messages(
        self,
        page: int = 1,
        page_size: int = 50,
        message_type: MessageType | None = None,
        since: str | None = None,
        unread_only: bool = False,
    ) -> PaginatedResponse[AgentMessage]:
        """Get messages for the current agent.

        Args:
            page: Page number
            page_size: Items per page
            message_type: Filter by message type
            since: Get messages since timestamp
            unread_only: Only unread messages

        Returns:
            Paginated list of messages
        """
        if not self._agent_id:
            raise ValueError("Agent not registered")

        params: dict[str, Any] = {"page": page, "pageSize": page_size}
        if message_type:
            params["type"] = message_type.value
        if since:
            params["since"] = since
        if unread_only:
            params["unreadOnly"] = True

        data = await self._client.get(
            f"{self._base_path}/{self._agent_id}/messages",
            params=params,
        )
        response_data = data.get("data", data)
        return PaginatedResponse[AgentMessage](
            data=[AgentMessage.model_validate(m) for m in response_data.get("data", [])],
            total=response_data.get("total", 0),
            page=response_data.get("page", page),
            page_size=response_data.get("pageSize", page_size),
            total_pages=response_data.get("totalPages", 0),
        )

    async def acknowledge_message(self, message_id: str) -> None:
        """Acknowledge receipt of a message.

        Args:
            message_id: Message ID
        """
        if not self._agent_id:
            raise ValueError("Agent not registered")
        await self._client.post(f"{self._base_path}/{self._agent_id}/messages/{message_id}/ack")
